### ISSUE TITLE
Incorrect parsing of default value in variable expansion

### ISSUE THREAD

**Issue by martin-schwenke** [koalaman/shellcheck#950]
#### For bugs
- Results in SC2031, but the problem isn't really with the rule
- 0.4.6
- I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/sh
b=$(d=$(dirname "$0") ; cd -P "$d" || exit ; dirname "$PWD")
bar="${FOOBAR:-${b}/foobar.d}"
echo "$bar"

```

#### Here's what shellcheck currently says:

In foo.sh line 2:
b=$(d=$(dirname "$0") ; cd -P "$d" || exit ; dirname "$PWD")
    ^-- SC2030: Modification of d is local (to subshell caused by $(..) expansion).

In foo.sh line 3:
bar="${FOOBAR:-${b}/foobar.d}"
       ^-- SC2031: d was modified in a subshell. That change might be lost.

#### Here's what I wanted or expected to see:

Nothing!

"d" is erroneously parsed as a variable name when it is part of a default value string.  This also happens if 'd' is preceded by '-' or '+' but not '_'.  So it looks like the default value is not treated at a string.  It looks like the tail end of it is treated as an identifier.

Thanks...

**Comment by martin-schwenke** [comment]
This didn't happen in 0.4.4, so I've hand bisected and found the commit that appears to introduce the regression:

a90b6d1 Count b as a reference in ${a:b}

Here's the difference between that commit and the previous one:

```
$ git reset --hard a90b6d1
HEAD is now at a90b6d1 Count b as a reference in ${a:b}
$ cabal build
Building ShellCheck-0.4.4...
Preprocessing library ShellCheck-0.4.4...
[11 of 17] Compiling ShellCheck.ASTLib ( ShellCheck/ASTLib.hs, dist/build/ShellCheck/ASTLib.o )
[12 of 17] Compiling ShellCheck.Parser ( ShellCheck/Parser.hs, dist/build/ShellCheck/Parser.o )
[13 of 17] Compiling ShellCheck.AnalyzerLib ( ShellCheck/AnalyzerLib.hs, dist/build/ShellCheck/AnalyzerLib.o )
[14 of 17] Compiling ShellCheck.Analytics ( ShellCheck/Analytics.hs, dist/build/ShellCheck/Analytics.o )
[15 of 17] Compiling ShellCheck.Checks.Commands ( ShellCheck/Checks/Commands.hs, dist/build/ShellCheck/Checks/Commands.o )
[16 of 17] Compiling ShellCheck.Analyzer ( ShellCheck/Analyzer.hs, dist/build/ShellCheck/Analyzer.o ) [ShellCheck.Analytics changed]
[17 of 17] Compiling ShellCheck.Checker ( ShellCheck/Checker.hs, dist/build/ShellCheck/Checker.o ) [ShellCheck.Analyzer changed]
Preprocessing executable 'shellcheck' for ShellCheck-0.4.4...
[11 of 18] Compiling ShellCheck.ASTLib ( ShellCheck/ASTLib.hs, dist/build/shellcheck/shellcheck-tmp/ShellCheck/ASTLib.o )
[12 of 18] Compiling ShellCheck.Parser ( ShellCheck/Parser.hs, dist/build/shellcheck/shellcheck-tmp/ShellCheck/Parser.o )
[13 of 18] Compiling ShellCheck.AnalyzerLib ( ShellCheck/AnalyzerLib.hs, dist/build/shellcheck/shellcheck-tmp/ShellCheck/AnalyzerLib.o )
[14 of 18] Compiling ShellCheck.Analytics ( ShellCheck/Analytics.hs, dist/build/shellcheck/shellcheck-tmp/ShellCheck/Analytics.o )
[15 of 18] Compiling ShellCheck.Checks.Commands ( ShellCheck/Checks/Commands.hs, dist/build/shellcheck/shellcheck-tmp/ShellCheck/Checks/Commands.o )
[16 of 18] Compiling ShellCheck.Analyzer ( ShellCheck/Analyzer.hs, dist/build/shellcheck/shellcheck-tmp/ShellCheck/Analyzer.o ) [ShellCheck.Analytics changed]
[17 of 18] Compiling ShellCheck.Checker ( ShellCheck/Checker.hs, dist/build/shellcheck/shellcheck-tmp/ShellCheck/Checker.o ) [ShellCheck.Analyzer changed]
Linking dist/build/shellcheck/shellcheck ...
$ dist/build/shellcheck/shellcheck -#!/bin/sh
b=$(d=$(dirname "$0") ; cd -P "$d" || exit ; dirname "$PWD")
bar="${FOOBAR:-${b}/foobar.d}"
echo "$bar"

In - line 2:
b=$(d=$(dirname "$0") ; cd -P "$d" || exit ; dirname "$PWD")
    ^-- SC2030: Modification of d is local (to subshell caused by $(..) expansion).


In - line 3:
bar="${FOOBAR:-${b}/foobar.d}"
       ^-- SC2031: d was modified in a subshell. That change might be lost.

$ git reset --hard 5a46eeb
HEAD is now at 5a46eeb Allow `#inline comments` without SC2046.
$ cabal build
Building ShellCheck-0.4.4...
Preprocessing library ShellCheck-0.4.4...
[13 of 17] Compiling ShellCheck.AnalyzerLib ( ShellCheck/AnalyzerLib.hs, dist/build/ShellCheck/AnalyzerLib.o )
[14 of 17] Compiling ShellCheck.Analytics ( ShellCheck/Analytics.hs, dist/build/ShellCheck/Analytics.o )
[15 of 17] Compiling ShellCheck.Checks.Commands ( ShellCheck/Checks/Commands.hs, dist/build/ShellCheck/Checks/Commands.o ) [ShellCheck.AnalyzerLib changed]
[16 of 17] Compiling ShellCheck.Analyzer ( ShellCheck/Analyzer.hs, dist/build/ShellCheck/Analyzer.o ) [ShellCheck.AnalyzerLib changed]
[17 of 17] Compiling ShellCheck.Checker ( ShellCheck/Checker.hs, dist/build/ShellCheck/Checker.o ) [TH]
Preprocessing executable 'shellcheck' for ShellCheck-0.4.4...
[13 of 18] Compiling ShellCheck.AnalyzerLib ( ShellCheck/AnalyzerLib.hs, dist/build/shellcheck/shellcheck-tmp/ShellCheck/AnalyzerLib.o )
[14 of 18] Compiling ShellCheck.Analytics ( ShellCheck/Analytics.hs, dist/build/shellcheck/shellcheck-tmp/ShellCheck/Analytics.o )
[15 of 18] Compiling ShellCheck.Checks.Commands ( ShellCheck/Checks/Commands.hs, dist/build/shellcheck/shellcheck-tmp/ShellCheck/Checks/Commands.o ) [ShellCheck.AnalyzerLib changed]
[16 of 18] Compiling ShellCheck.Analyzer ( ShellCheck/Analyzer.hs, dist/build/shellcheck/shellcheck-tmp/ShellCheck/Analyzer.o ) [ShellCheck.AnalyzerLib changed]
[17 of 18] Compiling ShellCheck.Checker ( ShellCheck/Checker.hs, dist/build/shellcheck/shellcheck-tmp/ShellCheck/Checker.o ) [TH]
Linking dist/build/shellcheck/shellcheck ...
$ dist/build/shellcheck/shellcheck -
#!/bin/sh
b=$(d=$(dirname "$0") ; cd -P "$d" || exit ; dirname "$PWD")
bar="${FOOBAR:-${b}/foobar.d}"
echo "$bar"
$
```

**Comment by martin-schwenke** [comment]
OK, this seems to fix it:

```
diff --git a/ShellCheck/AnalyzerLib.hs b/ShellCheck/AnalyzerLib.hs
index e9c0a1e..4e23b48 100644
--- a/ShellCheck/AnalyzerLib.hs
+++ b/ShellCheck/AnalyzerLib.hs
@@ -479,7 +479,7 @@ getOffsetReferences mods = fromMaybe [] $ do
     offsets <- match !!! 0
     return $ matchAllStrings variableNameRegex offsets
   where
-    re = mkRegex "^ *:(.*)"
+    re = mkRegex "^ *:([^-=?+].*)"
 
 getReferencedVariables parents t =
     case t of
```

I'll try to do the appropriate GitHub magic to do a merge request.

Note that, the substring indexing feature does not exist in POSIX shell.  It is an extension, in bash and others. I'm not sure how the parsing and shell dialects interact in ShellCheck but it seems a shame that a bug in that code is causing a false positive in POSIX shell.

**Comment by zapashcanon** [comment]
@koalaman, this can be closed. ;)

**Comment by martin-schwenke** [comment]
Just wondering if the policy is to leave issues open until there is a release?

**Comment by Nightfirecat** [comment]
As far as I've seen, this repository (like most) subscribes to the idea that issues are typically only opened against development code unless it's an issue tied to releases, or requiring a new release. (an example from one of my PRs, #988 closed #953)

**Comment by martin-schwenke** [comment]
Thanks.  Fair enough...  :-)

**Comment by koalaman** [comment]
Just me being slow :P
